### PR TITLE
added flow statistics and flow log

### PR DESCRIPTION
**Pull request by igalmk**
the backend sides of the 3rd command and 5th command are finished,
also fixed bugs found within validation
also fixed flow status setting after execution

### CODE REVIEW

File: StepperEngine/src/Steps/Step.java [hunk 1]
```diff
@@ -13,9 +13,9 @@ public abstract class  Step {
     private String alias;
     private Boolean hasAlias;
     private Boolean isReadOnly; // a readonly step doesn't change anything in the system.
-    private static double durationAvgInMs = 0.0;
-    private static double runTimeInMs = 0;
-    private static int stepRunsCounter = 0;
+    protected   double durationAvgInMs = 0.0;
+    protected   double runTimeInMs = 0;
+    protected   int startUpCounter = 0;
```
Comment (ofirlevy1) [inline StepperEngine/src/Steps/Step.java:18]: why do we need those fields? can't we manage with just the static fields?

File: StepperEngine/src/Steps/CollectFilesInFolderStep.java [hunk 2]
```diff
@@ -129,4 +131,17 @@ int addMatchingFilesToOutput(File[] files)
         return matchingFiles.getData().size();
     }
 
+    @Override
+    protected void updateStaticTimers() {
```
Comment (ofirlevy1) [inline StepperEngine/src/Steps/CollectFilesInFolderStep.java:135]: can't this logic be implemented in the "Step" parent class? (same question for the 2 below methods as well)

File: StepperEngine/src/Flow/Flow.java [hunk 3]
```diff
@@ -309,13 +313,19 @@ public void execute(){
...
     }
 
-    private void setFlowStatus(Step step){
-        if(step.getStatus()== Step.Status.Success)
-            status=Status.SUCCESS;
-        else if(step.getStatus()== Step.Status.Warning)
-            status=Status.WARNING;
-        else
+    private void setFlowStatus(){
+        if(steps.get(steps.size()-1).getStatus()== Step.Status.Failure){
```
Comment (ofirlevy1) [inline StepperEngine/src/Flow/Flow.java:317]: If step number x fails (and it's **blocking**), but there are x+1 steps? 

step number x+1 will not have fail status, no? (because it will not even be executed)

and then the "if" will not happen

Comment (igalmk) [reply StepperEngine/src/Flow/Flow.java:317]: if step number x fails and is blocking, an exception will be thrown and setFlowStatus will not be called, instead in the catch block the status will be set to failure.
if(steps.get(steps.size()-1).getStatus()== Step.Status.Failure), is here in case the last step failed and was non blocking
(i didn't see it as an user input error that needed to be validated)

File: StepperEngine/src/Steps/CollectFilesInFolderStep.java (hunk 2 above)
Comment (igalmk) [reply StepperEngine/src/Steps/CollectFilesInFolderStep.java:135]: the ipplementation cannot be Step class because the fields updated in the method aren't Step's fields, they are his children's
fields, and thus he will not recognize those fields

File: StepperEngine/src/Flow/FlowStatistics.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package Flow;
+
+import Steps.StepStatistics;
+
+import java.util.ArrayList;
+
+public class FlowStatistics {
```
[the remark this answers is not in the document]
Comment (igalmk) [reply StepperEngine/src/Flow/FlowStatistics.java:3]: FlowStatistics and also StepStatistics are more used as data transfers to the higher level, than logical objects.
i could store those fields in FlowStatistics as there will be only one instance of each flow, but i wanted FlowStatistics to be the same as StepStatistics aesthetic wise.

File: StepperEngine/src/Steps/Step.java (hunk 1 above)
Comment (igalmk) [reply StepperEngine/src/Steps/Step.java:18]: i have searched on the web of parent class having static fiields, and his children using those fields and found this:

https://stackoverflow.com/questions/17172033/java-and-inherited-static-members

tldr; if for example we run the FileDumperStep once and run FilesReanamerStep once, we would think that both start up counter would be 1, but in actuality it will be 2.

thus, we can't store the static members in Step, but in his children
